services/api/service.go

```diff
@@ -386,6 +403,57 @@ func (api *RelayAPI) startValidatorRegistrationDBProcessor() {
 	}
 }
 
+// simulateBlock sends a request for a block simulation to blockSimRateLimiter.
+func (api *RelayAPI) simulateBlock(opts blockSimOptions) error {
+	t := time.Now()
+	simErr := api.blockSimRateLimiter.send(opts.ctx, opts.req, opts.highPrio)
+	log := opts.log.WithFields(logrus.Fields{
+		"duration":   time.Since(t).Seconds(),
+		"numWaiting": api.blockSimRateLimiter.currentCounter(),
+	})
+	if simErr != nil {
+		log.WithError(simErr).Error("block validation failed")
+		return simErr
+	}
+	log.Info("block validation successful")
+	return nil
+}
+
+// processOptimisticBlock is called on a new goroutine when a optimistic block
+// needs to be simulated.
+func (api *RelayAPI) processOptimisticBlock(opts blockSimOptions) {
+	api.optimisticBlocksInFlight += 1
+	defer func() { api.optimisticBlocksInFlight -= 1 }()
+	api.optimisticBlocks.Add(1)
+	defer api.optimisticBlocks.Done()
+
+	builderPubkey := opts.req.Message.BuilderPubkey.String()
+	opts.log.WithFields(logrus.Fields{
+		"builderPubkey": builderPubkey,
+		// NOTE: this value is just an estimate because many goroutines could be
+		// updating api.optimisticBlocksInFlight concurrently. Since we just use
+		// it for logging, it is not atomic to avoid the performance impact.
+		"optBlocksInFlight": api.optimisticBlocksInFlight,
+	})
+
+	if err := api.simulateBlock(opts); err != nil {
+		api.log.WithError(err).Error("block simulation failed")
+		// Validation failed, demote all builders with the collateral id.
+		errs := api.datastore.SetBlockBuilderStatusByCollateralID(builderPubkey, common.OptimisticDemoted)
+		if len(errs) != 0 {
+			api.log.Error(fmt.Errorf("errors setting builder status by collateral id: %v", errs))
+		}
+
+		// Upsert into the builder demotion table but without the signed block
+		// or the signed registration, because we don't know if this bid will be
+		// accepted.
+		err = api.db.UpsertBuilderDemotion(&opts.req.BuilderSubmitBlockRequest, nil, nil)
```

Comment:
We should maybe use logs as a makeshift DB in case of an error here. We get that for free when we abstract away the logic in a separate function (as previously commented).